api/http/internal.go

```diff
@@ -76,11 +76,77 @@ func (h *InternalHandler) DecomissionDevice(c *gin.Context) {
 	})
 	err := h.app.DeleteIOTHubDevice(ctx, deviceID)
 	switch errors.Cause(err) {
-	case nil:
-		c.Status(http.StatusAccepted)
-	case app.ErrNoConnectionString:
+	case nil, app.ErrNoConnectionString:
 		c.Status(http.StatusNoContent)
 	default:
 		rest.RenderError(c, http.StatusInternalServerError, err)
 	}
 }
+
+type BulkResult struct {
+	Error bool       `json:"error"`
+	Items []BulkItem `json:"items"`
+}
+
+type BulkItem struct {
+	// Status code for the operation (translates to HTTP status)
+	Status int `json:"status"`
+	// Description in case of error
+	Description string `json:"description,omitempty"`
+	// Parameters used for producing BulkItem
+	Parameters map[string]interface{} `json:"parameters"`
+}
+
+const (
+	maxBulkItems = 100
+)
+
+// PUT /tenants/:tenant_id/bulk/devices/status
+func (h *InternalHandler) BulkSetDeviceStatus(c *gin.Context) {
+	var schema struct {
+		DeviceIDs []string   `json:"device_ids"`
+		Status    app.Status `json:"status"`
+	}
+	if err := c.ShouldBindJSON(&schema); err != nil {
+		rest.RenderError(c,
+			http.StatusBadRequest,
+			errors.Wrap(err, "invalid request body"),
+		)
+		return
+	} else if len(schema.DeviceIDs) > maxBulkItems {
+		rest.RenderError(c,
+			http.StatusBadRequest,
+			errors.New("too many bulk items: max 100 items per request"),
+		)
+		return
+	}
+	ctx := identity.WithContext(
+		c.Request.Context(),
+		&identity.Identity{
+			Tenant: c.Param("tenant_id"),
+		},
+	)
+	res := BulkResult{
+		Error: false,
+		Items: make([]BulkItem, len(schema.DeviceIDs)),
+	}
+	for i, id := range schema.DeviceIDs {
+		res.Items[i].Parameters = map[string]interface{}{
+			"device_id": schema.DeviceIDs[i],
+		}
+		err := h.app.SetDeviceStatus(ctx, id, schema.Status)
```

Comment:
NOTE: we are calling external service here (event twice in most cases), so, potentially, it's a very expensive operation;
it's OK for now (especially with the limit), but we should find a better way to do bulk operations in the future

Reply:
Sadly there's no bulk endpoint on the iot hub side for updating Device Twins (although they have one for [creating devices](https://docs.microsoft.com/en-us/rest/api/iothub/service/bulk-registry/update-registry)). I limited the maximum number of bulk items to 100 as a simple "work around" - even though the workflow is never called with more than one device.